src/dom-to-image-more.js

```diff
@@ -257,24 +258,38 @@
         return Promise.resolve(node)
             .then(makeNodeCopy)
             .then(function (clone) {
-                return cloneChildren(node, clone);
+                return cloneChildren(util.hasOpenShadowRoot(node) ? node.shadowRoot : node, clone);
             })
             .then(function (clone) {
                 return processClone(node, clone);
             });
 
         function makeNodeCopy(original) {
-            return util.isHTMLCanvasElement(original)
-                ? util.makeImage(original.toDataURL())
-                : original.cloneNode(false);
+            if (util.isHTMLCanvasElement(original)) {
+                return util.makeImage(original.toDataURL());
+            }
+            if (util.isInShadowRoot(original)) {
+                return cloneShadowNode(original);
+            }
+            return original.cloneNode(false);
+        }
+
+        function cloneShadowNode(original) {
+            const nextSibling = original.nextSibling;
+            const parentNode = original.parentNode;
+            // Bypass element encapsulation, by detaching from shadow node then restoring.
+            parentNode.removeChild(original);
+            const clone = original.cloneNode(false);
+            parentNode.insertBefore(original, nextSibling);
+            return clone;
         }
 
         function cloneChildren(original, clone) {
-            const originalChildren = original.childNodes;
+            const originalChildren = util.isShadowSlotElement(original) ? original.assignedNodes() : original.childNodes;
             let done = Promise.resolve();
 
             if (originalChildren.length !== 0) {
-                const originalComputedStyles = getComputedStyle(original);
+                const originalComputedStyles = getComputedStyle(util.isShadowRoot(original) ? original.host : original);
```

Comment:
Is it simply enough to do `original.host || original`? I'm really not familiar with the shadow DOM semantics as a whole, but this seems like a lot of fragile checks. I would welcome a chat about these changes.

Reply:
If `original` is a shadow DOM root (due to the ternary at line 261), we need to pass through the parent styling of its root element.

Let's take `<relative-time>` as the example.
In line 261 we "jump" from `<relative-time>` to its `#shadow-root` so we can copy its children (the text "X days ago"). The `<relative-time>` element can be accessed from `child=TextNode` by `child.parentNode.host`.

For maintainability I will put in extra variables (ternary output as a function call param is not that great). It's probably going to need explanation anyway so a one-liner comment would be good.

Reply:
Resolved in d656297